### PR TITLE
fix: concurrency bug when doing high volume pull query over java client

### PR DESCRIPTION
**Pull request by lucasbru**
### Description 

Fix for a race inside vert.x’s `RecordParser`. Two threads are hitting `RecordParserImpl.handleParsing`, one is entering from upstream via `RecordParser.handle`, and one thread is coming from the downstream via `ReadStream.resume`. Once our `PullQueryWriteStream` is half empty, we call the drain handler, which calls `RecordParser.resume`, which synchronously starts parsing new records from the buffer. `RecordParser.handleParsing` is not thread-safe.

The cause is that the monitor inside `PullQueryWriteStream` is insufficient to protect from races inside the `RecordParser`. This fix uses a separate monitor (the object monitor of the HTTP response object) to prevent races inside the `RecordParser`. To implement the fix, we define a shim wrapper around `RecordParser`, which makes sure to acquire the monitor both on `handle` and on `resume`.

In the future, we may want to fix this by changing the way the drain handler is called.

### Testing done 

No behavior changes in existing integration tests. Tested locally.

### Reviewer checklist
- [x] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [x] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")
- [x] Do these changes have compatibility implications for rollback? If so, ensure that the ksql [command version](https://github.com/confluentinc/ksql/blob/master/ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/Command.java#L41) is bumped.

### CODE REVIEW

File: ksqldb-engine/src/main/java/io/confluent/ksql/query/PullQueryWriteStream.java [hunk 1]
```diff
@@ -208,12 +208,8 @@ private PullQueryRow pollRow() {
 
     polled.handler.handle(new SucceededFuture<>(null, null));
 
-    if (monitor.enterIf(atHalfCapacity)) {
-      try {
-        drainHandler.forEach(h -> h.handle(null));
-      } finally {
-        monitor.leave();
-      }
+    if (isDone() || size() <= queueCapacity / 2) {
```
Comment (cadonna) [inline ksqldb-engine/src/main/java/io/confluent/ksql/query/PullQueryWriteStream.java:211]: Why did you remove the monitor guard?

File: ksqldb-engine/src/main/java/io/confluent/ksql/query/PullQueryWriteStream.java [hunk 2]
```diff
@@ -208,12 +208,8 @@ private PullQueryRow pollRow() {
...
+      drainHandler.forEach(h -> h.handle(null));
     }
```
Comment (cadonna) [inline ksqldb-engine/src/main/java/io/confluent/ksql/query/PullQueryWriteStream.java:213]: I think the drain handler is called too often also in the original code. As far as I understand the drain handler should be called to resume the source that is piped into the write stream (i.e. the http response) after is has been paused. Just because the size of the queue is less than half its capacity, it does not mean the http response has been paused, it could also be that the response gets content slower.
During our investigation, I tried to call the drain handler only when the http response has been paused before and that has already reduced the error rate significantly but did not solve the issue, though.
What do you think of also reducing the calls to the drain handler?
Probably, we should not do it in this PR since for testing your fix we want to hit the issue often. Maybe let's consider it an optimization that we might or might not add afterwards.

File: ksqldb-rest-client/src/main/java/io/confluent/ksql/rest/client/SyncronizedRecordParser.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
+package io.confluent.ksql.rest.client;
```
Comment (cadonna) [inline ksqldb-rest-client/src/main/java/io/confluent/ksql/rest/client/SyncronizedRecordParser.java:1]: Don't you need a license header here?

**Review by cadonna — COMMENTED** [review]
@lucasbru Thanks for the fix!

Here my feedback!

File: ksqldb-engine/src/main/java/io/confluent/ksql/query/PullQueryWriteStream.java (hunk 2 above)
Comment (lucasbru) [reply ksqldb-engine/src/main/java/io/confluent/ksql/query/PullQueryWriteStream.java:213]: Yes, I think the whole way the drain handler is used is quite mysterious to me. It's called too often, it's protected by a monitor whose purpose is not clear, it's called from a different thread (other WriteStreams call it from `write`, so basically from the same context).

I think you are right that it should be called less often, however this change is really to do the minimum change necessary. It's an idea to try this in a separate PR, but then again, not sure if there is enough interest to improve pull queries at this point.

File: ksqldb-engine/src/main/java/io/confluent/ksql/query/PullQueryWriteStream.java (hunk 1 above)
Comment (lucasbru) [reply ksqldb-engine/src/main/java/io/confluent/ksql/query/PullQueryWriteStream.java:211]: Firstly, it's not clear that we need it, because the drain handler doesn't modify the state of the WriteStream - it only modifies the state of the RecordParser, and the RecordParser now has a separate monitor.

More importantly, the monitor now can cause a deadlock:
 - `RecordParser.handle` acquires the new monitor that I introduced, and subsequently calls `PullQueryWriteStream.handle`, which attempts to acquire the `monitor` in this context.
 - `PullQueryWriteStream.pollRow` acquires `monitor` and calls `RecordParser.resume` which attempts to acquire the monitor that I introduced.

So we remove aquiring `monitor` here.

File: ksqldb-rest-client/src/main/java/io/confluent/ksql/rest/client/SyncronizedRecordParser.java (hunk 3 above)
Comment (lucasbru) [reply ksqldb-rest-client/src/main/java/io/confluent/ksql/rest/client/SyncronizedRecordParser.java:1]: true, done

File: ksqldb-rest-client/src/main/java/io/confluent/ksql/rest/client/SyncronizedRecordParser.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+import io.vertx.codegen.annotations.Fluent;
+import io.vertx.codegen.annotations.Nullable;
+import io.vertx.core.Handler;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.parsetools.RecordParser;
+import io.vertx.core.streams.ReadStream;
+
+@SuppressWarnings("unused")
+public class SyncronizedRecordParser implements RecordParser {
```
Comment (cadonna) [inline ksqldb-rest-client/src/main/java/io/confluent/ksql/rest/client/SyncronizedRecordParser.java:26]: typo:
`SyncronizedRecordParser` -> `SynchronizedRecordParser`

**Comment by cadonna** [comment]
There are some build errors.

File: ksqldb-rest-client/src/main/java/io/confluent/ksql/rest/client/SyncronizedRecordParser.java (hunk 4 above)
Comment (lucasbru) [reply ksqldb-rest-client/src/main/java/io/confluent/ksql/rest/client/SyncronizedRecordParser.java]: 😞

**Comment by AlanConfluent** [comment]
> Fix for a race inside vert.x’s RecordParser. Two threads are hitting RecordParserImpl.handleParsing, one is entering from upstream via RecordParser.handle, and one thread is coming from the downstream via ReadStream.resume. Once our PullQueryWriteStream is half empty, we call the drain handler, which calls RecordParser.resume, which synchronously starts parsing new records from the buffer. RecordParser.handleParsing is not thread-safe.

This is an interesting insight.  It occurs to me that many of these pieces like `RecordParser` assume they are being called from the same vertx thread as both the things producing to it and consuming from it.  If you look at the callback within the http response hander, they do exactly this: https://github.com/eclipse-vertx/vert.x/blob/master/src/main/java/io/vertx/core/http/impl/Http1xServerResponse.java#L542

I think we've broken that model here, as you've pointed out.  There's the Vertx thread thread which is writing to the RecordParser, and then the Pull query thread which is pulling from the queue (and invoking the drain callback and calling resume).  One solution would be to just invoke the handler using the same Vertx thread which registers the drain handler.  That might be simpler than adding synchronization.

File: ksqldb-rest-client/src/main/java/io/confluent/ksql/rest/client/SynchronizedRecordParser.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+  }
+
+  @Override
+  public RecordParser fetch(final long l) {
+    return delegate.fetch(l);
+  }
+
+  @Override
+  public RecordParser resume() {
+    synchronized (source) {
```
[the remark this answers is not in the document]
Comment (AlanConfluent) [reply ksqldb-rest-client/src/main/java/io/confluent/ksql/rest/client/SynchronizedRecordParser.java:146]: Also, I'm slightly worried about other internal uses of state in `RecordParser` that could access state concurrently with these synchronized calls. It seems like you would have to audit the delegate implementation to know which methods require additional synchronization and potentially add them here.

File: ksqldb-engine/src/main/java/io/confluent/ksql/query/PullQueryWriteStream.java (hunk 1 above)
Comment (lucasbru) [reply ksqldb-engine/src/main/java/io/confluent/ksql/query/PullQueryWriteStream.java:211]: @cadonna and I tried running it on the context of the Read Stream, but failed to fix it this way. The problem was that the pull queries would not make progress, and I couldn't determine what was causing it. Today I gave this another look, and I was able to figure it our -- we are calling the drain handler wayyyy to often, so we are clogging the netty event loop completely with drain handler calls. The pull queries make progress, but the netty event loops are essentially at 100% CPU just going through drain handler calls. I think the correct fix is to call the drain handler only exactly once - the pipe implementation will re-register its drain handler when the `WriteStream` reaches capacity. @AlanConfluent could you have a look at the alternative fix here: https://github.com/confluentinc/ksql/pull/10077
